core/circuit.go

```diff
@@ -0,0 +1,244 @@
+	maxCurrent    float64           // max allowed current
+	parentCircuit *Circuit          // parent circuit reference, used to determine current limits from hierarchy
+	phaseMeter    api.PhaseCurrents // meter to determine phase current
+}
+
+// GetCurrent determines current in use. Implements consumer interface
+// TBD: phase perfect evaluation
+func (circuit *Circuit) MaxPhasesCurrent() (float64, error) {
+	var current float64
+
+	i1, i2, i3, err := circuit.phaseMeter.Currents()
+	if err != nil {
+		return 0, fmt.Errorf("failed getting meter currents: %w", err)
+	}
+	circuit.log.DEBUG.Printf("(%s) meter currents: %.3gA", circuit.Title, []float64{i1, i2, i3})
+	circuit.publish("meterCurrents", []float64{i1, i2, i3})
+	// TODO: phase adjusted handling. Currently we take highest current from all phases
+	current = math.Max(math.Max(i1, i2), i3)
+
+	circuit.log.DEBUG.Printf("(%s) actual current: %.1fA", circuit.Title, current)
+	circuit.publish("actualCurrent", current)
+	return current, nil
+}
+
+// GetRemainingCurrent avaialble current based on limit and consumption
+// checks down up to top level parent
+func (circuit *Circuit) GetRemainingCurrent() float64 {
+	circuit.log.TRACE.Printf("(%s) get available current", circuit.Title)
+	// first update current current, mainly to regularly publish the value
+	current, err := circuit.MaxPhasesCurrent()
+	if err != nil {
+		circuit.log.ERROR.Printf("(%s) max phase currents: %v", circuit.Title, err)
+		return 0
+	}
+	curAvailable := circuit.maxCurrent - current
+	if curAvailable < 0.0 {
+		circuit.log.WARN.Printf("(%s) overload detected (%s) - currents: %.1fA, allowed max current is: %.1fA\n", circuit.Title, circuit.Title, current, circuit.maxCurrent)
+		circuit.publish("overload", true)
+	} else {
+		circuit.publish("overload", false)
+	}
+	// check parent circuit, return lowest
+	if circuit.parentCircuit != nil {
+		circuit.log.TRACE.Printf("(%s) get available current from parent: %s", circuit.Title, circuit.parentCircuit.Title)
+		curAvailable = math.Min(curAvailable, circuit.parentCircuit.GetRemainingCurrent())
+	}
+	circuit.log.DEBUG.Printf("(%s) circuit using %.1fA, %.1fA available", circuit.Title, current, curAvailable)
+	return curAvailable
+}
+
+// NewCircuit a circuit with defaults
+func NewCircuit(t string, limit float64, pm api.PhaseCurrents, l *util.Logger) *Circuit {
+	circuit := &Circuit{
+		log:        l,
+		Title:      t,
+		maxCurrent: limit,
+		phaseMeter: pm,
+	}
+	circuitId += 1
+	return circuit
+}
+
+// NewCircuitFromConfig creates circuit from config
+// using site to get access to the grid meter if configured, see cp.Meter() for details
+// returns a map of circuit name and circuit ref
+func NewCircuitFromConfig(cp configProvider, other map[string]interface{}) (map[string]*Circuit, map[string]*VMeter, error) {
+	var circuitCfg = new(CircuitConfig)
+
+	if err := util.DecodeOther(other, circuitCfg); err != nil {
+		return nil, nil, err
+	}
+	if circuitCfg.Name == "" {
```

Comment:
Circuit sollte seinen Namen nicht kennen

Reply:
Erzeugung analog:

	cp.chargers = make(map[string]api.Charger)
	for id, cc := range conf.Chargers {
		if cc.Name == "" {
			return fmt.Errorf("cannot create %s charger: missing name", humanize.Ordinal(id+1))
		}